beacon-chain/state/types.go

```diff
@@ -103,15 +104,25 @@ type ReadOnlyValidator struct {
 	validator *ethpb.Validator
 }
 
+func (r *reference) Refs() uint {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.refs
+}
+
 func (r *reference) AddRef() {
+	r.lock.Lock()
 	r.refs++
+	r.lock.Unlock()
 }
 
 func (r *reference) MinusRef() {
+	r.lock.Lock()
 	// Do not reduce further if object
 	// already has 0 reference to prevent overflow.
 	if r.refs == 0 {
 		return
 	}
 	r.refs--
+	r.lock.Unlock()
```

Comment:
this needs to be deferred, otherwise we hold the lock indefinitely, if we return early

Reply:
Yep, oops!